tsz-compress/src/v2/encode.rs

```diff
@@ -0,0 +1,369 @@
+        let mut fits = [true; 6];
+
+        // Can not emit with any case of delta compression if queue is empty
+        if self.is_empty() {
+            return 0;
+        }
+
+        // Can not emit with case v of delta compression if number of samples < 10
+        if self.len() < 10 {
+            fits[0] = false;
+        }
+
+        // Can not emit with case iv of delta compression if number of samples < 5.
+        if self.len() < 5 {
+            fits[1] = false;
+        }
+
+        // Can not emit with case iii of delta compression if number of samples < 4
+        if self.len() < 4 {
+            fits[2] = false;
+        }
+
+        // Can not emit with case ii of delta compression if number of samples < 3
+        if self.len() < 3 {
+            fits[3] = false;
+        }
+
+        // Can not emit with case ii of delta compression if number of samples < 2
+        if self.len() < 2 {
+            fits[4] = false;
+        }
+
+        // Check if the values will fit in the cases
+        let values = self.peak_bitcounts::<10>();
+        for (index, bits_required) in values.into_iter().enumerate() {
+            if (index < 2) & (bits_required > 16) {
+                fits[4] = false;
+            }
+            if (index < 3) & (bits_required > 10) {
+                fits[3] = false;
+            }
+            if (index < 4) & (bits_required > 8) {
+                fits[2] = false;
+            }
+            if (index < 5) & (bits_required > 6) {
+                fits[1] = false;
+            }
+            if (index < 10) & (bits_required > 3) {
+                fits[0] = false;
+            }
+        }
+
+        // Emit according to priority of cases
+        if fits[0] {
+            push_three_bits(self, out);
+            return 10;
+        } else if fits[1] {
+            push_six_bits(self, out);
+            return 5;
+        } else if fits[2] {
+            push_eight_bits(self, out);
+            return 4;
+        } else if fits[3] {
+            push_ten_bits(self, out);
+            return 3;
+        } else if fits[4] {
+            push_sixteen_bits(self, out);
+            return 2;
+        } else if fits[5] {
+            unsafe {
+                push_thirty_two_bits(self, out);
+            }
+            return 1;
+        }
+        0
+    }
+}
+
+// Delta-Delta Encoding
+pub trait EmitDeltaDeltaBits {
+    /// Emits bits according to the most efficient case of Delta Compression.
+    /// Returns the number of elements popped from the queue.
+    fn emit_delta_delta_bits(&mut self, out: &mut HalfVec) -> usize;
+}
+
+fn emit_popped_values<const N: usize>(
+    bitcounts: &[usize; N],
+    values: &[usize; N],
+    out: &mut HalfVec,
+) {
+    for (bits, value) in bitcounts.into_iter().zip(values.into_iter()) {
+        match bits {
+            0 => out.push(HalfWord::Half(0b0000)),
+            // -1 => out.push(HalfWord::Half(0b0001)),
```

Comment:
Having this disabled will hit the _ case

Reply:
We are pushing the zigzag values of data into the CompressionQueue, so it is not possible to have negative values popped out of a CompressionQueue.
```rust
unsafe fn write<T: Bits + Sized>(&mut self, index: usize, t: T) {
     let (zbits, zcount) = t.zigzag_bits();
     *self.zigzag.get_unchecked_mut(index) = zbits;
     *self.bitcount.get_unchecked_mut(index) = zcount;
}
```